pkg/payerreport/verifier.go

```diff
@@ -54,19 +66,181 @@ func (p *PayerReportVerifier) IsValidReport(
 		zap.String("new_report_id", newReportID.String()),
 	)
 
-	if err := validateReportTransition(prevReport, newReport); err != nil {
+	if err = validateReportTransition(prevReport, newReport); err != nil {
 		log.Warn("invalid report transition", zap.Error(err))
 		return false, nil
 	}
 
-	if err := validateReportStructure(newReport); err != nil {
+	if err = validateReportStructure(newReport); err != nil {
 		log.Warn("invalid report content", zap.Error(err))
 		return false, nil
 	}
 
+	var isValidMerkleRoot bool
+	// If the start and end sequence IDs are the same, the report is empty and the merkle root must always be the hash of an empty set
+	if newReport.StartSequenceID == newReport.EndSequenceID {
+		// TODO:nm validate that the merkle root is the hash of an empty set
+		isValidMerkleRoot = true
+	} else {
+		if isValidMerkleRoot, err = p.verifyMerkleRoot(ctx, newReport); err != nil {
+			return false, err
+		}
+	}
+
+	return isValidMerkleRoot, nil
+}
+
+// Re-generates the payer map and verifies that the merkle root in the report matches the newly generated one
+func (p *PayerReportVerifier) verifyMerkleRoot(
+	ctx context.Context,
+	report *PayerReport,
+) (bool, error) {
+	startEnvelope, endEnvelope, err := p.getStartAndEndMessages(ctx, report)
+	if err != nil {
+		return false, err
+	}
+	// If the start sequence ID is 0, it is the first report and we should start from minute 0 since there are no preceding reports
+	var startMinute int32
+	if startEnvelope.OriginatorSequenceID() == 0 {
+		startMinute = 0
+	} else {
+		startMinute, err = getMinuteFromEnvelope(startEnvelope)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	endMinute, err := getMinuteFromEnvelope(endEnvelope)
+	if err != nil {
+		return false, err
+	}
+
+	// Validate the report ends after it starts
+	if startMinute > endMinute {
+		return false, ErrInvalidReportStart
+	}
```

Comment:
nit: I wonder if this can ever happen as we're checking start sequence id > end sequence id. Good to have this check anyway.